demos/programs/getsubres/getsubres.man
.\" $XConsortium: getsubres.man /main/5 1996/04/22 23:15:06 pascale $
.\" Motif
.\"
.\" Copyright (c) 1987-2012, The Open Group. All rights reserved.
.\"
.\" These libraries and programs are free software; you can
.\" redistribute them and/or modify them under the terms of the GNU
.\" Lesser General Public License as published by the Free Software
.\" Foundation; either version 2.1 of the License, or (at your option)
.\" any later version.
.\"
.\" These libraries and programs are distributed in the hope that
.\" they will be useful, but WITHOUT ANY WARRANTY; without even the
.\" implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
.\" PURPOSE. See the GNU Lesser General Public License for more
.\" details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with these librararies and programs; if not, write
.\" to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
.\" Floor, Boston, MA 02110-1301 USA
...\" 
...\" 
...\" HISTORY
.TH getsubres 1X MOTIF "Demonstration programs"
.SH NAME
\*Lgetsubres\*O\ - a viewer of Motif widget resources
.SH SYNOPSIS
.sS
\*Lgetsubres\*O [All|<classname>]
.sE
.SH DESCRIPTION
\*LGetsubres\*O
uses XmGetSecondaryResourceData to fetch the widget
resources that XtGetResourceList won't see.

You can specify which widget resources to display on the command
line by using the syntax "getsubres <name>" with one of those names:

  Object, RectObj, Widget, Core, Composite, Constraint,
  Primitive, ScrollBar, ArrowB, List, Separator, TextF,
  Label, DrawnB, PushB, ToggleB, CascadeB, Text,
  Gadget, ArrowBG, SeparatorG, LabelG, PushBG, ToggleBG, 
  CascadeBG, Manager, BulletinB, DrawingA, Frame, MainW, 
  Form, RowCol, ScrolledW, PanedW, SelectionB, FileSB, 
  MessageB, Scale, Command, Display, Screen, DragContext, 	
  DragIcon, DropSiteMgr, DropTransfer, DragOverS,
  Shell, OverrideShell, WMShell, MenuShell, VendorS,
  ToplevelShell, ApplicationShell, TransientShell, 
  DialogShell, Container, IconGadget, Notebook, ComboBox, 
  SpinBox. (default is Widget)

Or you can use "getsubres All".

.SH X Defaults
The class of this application is \fIXmdGetsubres\fR.  

.SH Editres: 
Getsubres is Editres aware, so you can use 'editres' to see the
names of the widgets in the instance hierarchy.



.attic/ChangeLog
2017-11-10 11:49:10 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1665 (HEAD, origin/master, origin/HEAD, master)

2017-10-30 23:04:15 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1651

2017-10-30 22:22:31 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1650

2017-10-30 21:47:21 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1652

2017-08-31 06:28:27 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1655

2017-08-28 05:33:53 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1625

2017-08-26 06:12:14 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1656

2017-08-23 04:10:55 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1658

2017-08-22 06:05:00 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1659

2017-08-22 03:27:36 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1660

2017-03-28 01:40:33 +0300 Oleksiy Chernyavskyy 

	* updated version and release data

2017-03-28 01:32:30 +0300 Oleksiy Chernyavskyy 

	* removed redundant includes

2017-03-28 01:11:59 +0300 Oleksiy Chernyavskyy 

	* removed bugfix 1565. See details on http://bugs.motifzone.net/show_bug.cgi?id=1565

2016-11-09 07:19:53 +0200 Alex 

	* obsolete malloc.h changed on stdlib.h

2016-06-11 05:03:21 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-11 05:02:55 +0300 Oleksiy Chernyavskyy 

	* updated release 2.3.6 information

2016-06-11 04:57:54 +0300 Oleksiy Chernyavskyy 

	* fixid parallel building (make -j) error

2016-06-11 04:57:20 +0300 Oleksiy Chernyavskyy 

	* removed automatic configure start from autogen.sh

2016-06-11 04:26:48 +0300 Oleksiy Chernyavskyy 

	* minor fix

2016-06-11 04:21:53 +0300 Oleksiy Chernyavskyy 

	* removed extra dist Imakefile from all Makefile.am files

2016-06-11 04:06:37 +0300 Oleksiy Chernyavskyy 

	* Purified copyright headers (Motif is LGPL covered). Removed many Imakefile's and Imake-related stuff

2016-06-10 23:40:43 +0300 Oleksiy Chernyavskyy 

	* updated release notes and version 2.3.5 -> 2.3.6

2016-06-10 23:27:56 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-10 23:15:00 +0300 Oleksiy Chernyavskyy 

	* bug 1654 fix

2016-06-10 21:58:16 +0300 Oleksiy Chernyavskyy 

	* old patches. need review

2016-03-29 03:48:59 +0300 Oleksiy Chernyavskyy 

	* removed obsolete files

2016-03-17 04:23:17 +0200 Oleksiy Chernyavskyy 

	* release 2.3.5 (tag: release_2_3_5)

2016-03-16 23:07:12 +0200 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-03-16 23:06:24 +0200 Oleksiy Chernyavskyy 

	* added mkchangelog script for automated generation of the ChangeLog

2016-03-16 00:46:49 +0200 Oleksiy Chernyavskyy 

	* Reimplemented bugfix 1565

2016-02-26 03:36:51 +0200 Oleksiy Chernyavskyy 

	* Updated RPM creation instructions

2014-08-20 15:45:41 +0300 Oleksiy Chernyavskyy 

	* fixed bug 1636

2013-08-21 11:49:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1612 (Label size computed wrong within a Form).

2013-07-11 18:14:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1618 (Eliminates 29 warnings of some type in modules from "demos" catalog).

2013-07-11 15:20:54 +0300 Mykola Vshyvkov 

	* Fixed bug #1616 (Eliminates 23 warnings "suggest parentheses around assignment used" in modules from "demos" catalog).

2013-07-11 15:01:50 +0300 Mykola Vshyvkov 

	* Fixed bug #1617 (Eliminates 23 warnings "cast to pointer from integer of different size" in modules from "demos" catalog).

2013-07-11 12:35:30 +0300 Mykola Vshyvkov 

	* Fixed bug #1615 (Eliminates 35 warnings "enumeration value XXX not handled" in modules from "demos" catalog).

2013-07-10 06:55:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1614 (Eliminates 36 warnings "variable may be used uninitialized" in modules from "demos" catalog).

2013-07-10 05:05:28 +0300 Mykola Vshyvkov 

	* Fixed bug #1613 (Eliminates 60 warnings "implicit declaration of function" in modules from "demos" catalog).

2013-07-04 14:08:09 +0300 Mykola Vshyvkov 

	* Fixed bug #1611 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls in modules of "clients" catalog).

2013-07-03 21:27:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1610 (Eliminates 15 warnings of some types in modules of "clients" catalog).

2013-07-03 18:52:21 +0300 Mykola Vshyvkov 

	* Fixed bug #1609 (Eliminates 18 warnings  of "incompatible implicit declaration" in modules of "clients" catalog).

2013-07-03 14:58:22 +0300 Mykola Vshyvkov 

	* Fixed bug #1608 (Eliminates 30 warnings  of "array subscript has type 'char'" in modules of "clients" catalog).

2013-07-03 14:35:04 +0300 Mykola Vshyvkov 

	* Fixed bug #1607 (Eliminates 92 warnings  of "variable may be used uninitialized" in modules of "clients" catalog).

2013-07-03 12:48:55 +0300 Mykola Vshyvkov 

	* Fixed bug #1606 (Eliminates 130 warnings  of "expects argument of type" in modules of "clients" catalog).

2013-07-02 13:40:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 20:52:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 19:07:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1604 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls).

2013-07-01 19:02:25 +0300 Mykola Vshyvkov 

	* Fixed bug #1605 (Eliminates 14 warnings of some types in WML tool).

2013-06-26 23:45:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1603 (Eliminates 18 warnings of some types in some modules).

2013-06-26 19:21:20 +0300 Mykola Vshyvkov 

	* Fixed bug #1602 (Eliminates 6 warnings "implicit declaration of function" in some modules).

2013-06-26 18:36:47 +0300 Mykola Vshyvkov 

	* Fixed bug #1601 (Eliminates 6 warnings: "suggest explicit braces to avoid ambiguous" in some modules).

2013-06-26 17:40:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1600 (Eliminates 26 warnings about using uninitialized variables in some modules).

2013-06-26 15:34:53 +0300 Mykola Vshyvkov 

	* Fixed bug #1599 (Eliminates 14 warnings of incompatible type argument from some modules).

2013-06-25 17:37:02 +0300 Mykola Vshyvkov 

	* Updated ChangeLog.

2013-06-25 13:57:31 +0300 Mykola Vshyvkov 

	* Merge branch 'openmotif_2_3' of ssh://git.code.sf.net/p/motif/code into openmotif_2_3


demos/programs/getsubres/getsubres.c
/* $XConsortium: getsubres.c /main/6 1995/07/15 20:45:28 drk $ */
/**
 * Motif
 *
 * Copyright (c) 1987-2012, The Open Group. All rights reserved.
 *
 * These libraries and programs are free software; you can
 * redistribute them and/or modify them under the terms of the GNU
 * Lesser General Public License as published by the Free Software
 * Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * These libraries and programs are distributed in the hope that
 * they will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with these librararies and programs; if not, write
 * to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301 USA
 *
 */

#include <stdlib.h>
#include <Xm/XmAll.h>
#include <Xmd/Help.h>

/* This is no your ordinary demo, it displays widget class info and
   needs a couple of P.h file usually reserved for widget writers */
#include <Xm/VaSimpleP.h>

#include <Xm/SashP.h>
#include <Xm/TearOffBP.h>
#include <Xm/GrabShellP.h>
#include <Xm/IconHP.h>
#include <Xm/ClipWindowP.h>

static void QuitCB(Widget w, XtPointer client, XtPointer callback);
static void HelpCB(Widget w, XtPointer client, XtPointer callback);

typedef struct _ClassArrayRec {
    String name ;
    WidgetClass class ;
} ClassArrayRec ;


/* Cannot initialize everything in C */
static ClassArrayRec class_array[] = {
  { "Object",  (WidgetClass) NULL },
  { "RectObj",  (WidgetClass) NULL },
  { "Widget",  (WidgetClass) NULL },
  { "Core",  (WidgetClass) NULL },
  { "Composite",  (WidgetClass) NULL },
  { "Constraint",  (WidgetClass) NULL },

  { "Primitive",  (WidgetClass) NULL },
  { "ScrollBar",  (WidgetClass) NULL },
  { "ArrowB",  (WidgetClass) NULL },
  { "List",  (WidgetClass) NULL },
  { "Separator",  (WidgetClass) NULL },
  { "TextF",  (WidgetClass) NULL },
  { "Label",  (WidgetClass) NULL },
  { "DrawnB",  (WidgetClass) NULL },
  { "PushB",  (WidgetClass) NULL },
  { "ToggleB",  (WidgetClass) NULL },
  { "CascadeB",  (WidgetClass) NULL },

  { "Text",  (WidgetClass) NULL },

  { "Gadget",  (WidgetClass) NULL },
  { "ArrowBG",  (WidgetClass) NULL },
  { "SeparatorG",  (WidgetClass) NULL },
  { "LabelG",  (WidgetClass) NULL },
  { "PushBG",  (WidgetClass) NULL },
  { "ToggleBG", (WidgetClass) NULL },
  { "CascadeBG",  (WidgetClass) NULL },

  { "Manager", (WidgetClass) NULL },
  { "BulletinB", (WidgetClass) NULL },
  { "DrawingA", (WidgetClass) NULL },
  { "Frame", (WidgetClass) NULL  },
  { "MainW", (WidgetClass) NULL },
  { "Form", (WidgetClass) NULL },
  { "RowCol", (WidgetClass) NULL },
  { "ScrolledW", (WidgetClass) NULL },
  { "PanedW", (WidgetClass) NULL },
  { "SelectionB", (WidgetClass) NULL },
  { "FileSB", (WidgetClass) NULL },
  { "MessageB", (WidgetClass) NULL },
  { "Scale",  (WidgetClass) NULL },
  { "Command",  (WidgetClass) NULL },

  { "Shell",  (WidgetClass) NULL },
  { "OverrideShell",  (WidgetClass) NULL },
  { "WMShell",  (WidgetClass) NULL },

  { "MenuShell",  (WidgetClass) NULL },

  { "VendorS",  (WidgetClass) NULL },
  { "ToplevelShell",  (WidgetClass) NULL },
  { "ApplicationShell",  (WidgetClass) NULL },
  { "TransientShell",  (WidgetClass) NULL },

  { "DialogShell",  (WidgetClass) NULL },

  { "Sash",  (WidgetClass) NULL },
  { "TearOffB",  (WidgetClass) NULL },

  { "Display",  (WidgetClass) NULL },
  { "Screen",  (WidgetClass) NULL },
  { "DragContext",  (WidgetClass) NULL },
  { "DragIcon",  (WidgetClass) NULL },
  { "DropSiteMgr",  (WidgetClass) NULL },
  { "DropTransfer",  (WidgetClass) NULL },

  { "DragOverS",  (WidgetClass) NULL },

  { "GrabShell",  (WidgetClass) NULL },
  { "IconH",  (WidgetClass) NULL },
  { "ClipWindow",  (WidgetClass) NULL },

  { "Container",  (WidgetClass) NULL },
  { "IconG",  (WidgetClass) NULL },
  { "Notebook",  (WidgetClass) NULL },
  { "SpinBox",  (WidgetClass) NULL },
  { "SimpleSpinBox",  (WidgetClass) NULL },
  { "ComboBox",  (WidgetClass) NULL },
};

/* stupid c compiler */
static void InitClassArray () {
  int n;

  n = 0;
  class_array[n].class = objectClass; n++;
  class_array[n].class = rectObjClass; n++;
  class_array[n].class = widgetClass; n++;  /* unNamed */
  class_array[n].class = coreWidgetClass; n++;
  class_array[n].class = compositeWidgetClass; n++;
  class_array[n].class = constraintWidgetClass; n++;

  class_array[n].class = xmPrimitiveWidgetClass; n++;
  class_array[n].class = xmScrollBarWidgetClass; n++;
  class_array[n].class = xmArrowButtonWidgetClass; n++;
  class_array[n].class = xmListWidgetClass; n++;
  class_array[n].class = xmSeparatorWidgetClass; n++;
  class_array[n].class = xmTextFieldWidgetClass; n++;
  class_array[n].class = xmLabelWidgetClass; n++;
  class_array[n].class = xmDrawnButtonWidgetClass; n++;
  class_array[n].class = xmPushButtonWidgetClass; n++;
  class_array[n].class = xmToggleButtonWidgetClass; n++;
  class_array[n].class = xmCascadeButtonWidgetClass; n++;
  class_array[n].class = xmTextWidgetClass; n++;

  class_array[n].class = xmGadgetClass; n++;
  class_array[n].class = xmArrowButtonGadgetClass; n++;
  class_array[n].class = xmSeparatorGadgetClass; n++;
  class_array[n].class = xmLabelGadgetClass; n++;
  class_array[n].class = xmPushButtonGadgetClass; n++;
  class_array[n].class = xmToggleButtonGadgetClass; n++;
  class_array[n].class = xmCascadeButtonGadgetClass; n++;

  class_array[n].class = xmManagerWidgetClass; n++;
  class_array[n].class = xmBulletinBoardWidgetClass; n++;
  class_array[n].class = xmDrawingAreaWidgetClass; n++;
  class_array[n].class = xmFrameWidgetClass; n++;
  class_array[n].class = xmMainWindowWidgetClass; n++;
  class_array[n].class = xmFormWidgetClass; n++;
  class_array[n].class = xmRowColumnWidgetClass; n++;
  class_array[n].class = xmScrolledWindowWidgetClass; n++;
  class_array[n].class = xmPanedWindowWidgetClass; n++;
  class_array[n].class = xmSelectionBoxWidgetClass; n++;
  class_array[n].class = xmFileSelectionBoxWidgetClass; n++;
  class_array[n].class = xmMessageBoxWidgetClass; n++;
  class_array[n].class = xmScaleWidgetClass; n++;
  class_array[n].class = xmCommandWidgetClass; n++;

  class_array[n].class = shellWidgetClass; n++;
  class_array[n].class = overrideShellWidgetClass; n++;
  class_array[n].class = wmShellWidgetClass; n++;

  class_array[n].class = xmMenuShellWidgetClass; n++;

  class_array[n].class = vendorShellWidgetClass; n++;
  class_array[n].class = topLevelShellWidgetClass; n++;
  class_array[n].class = applicationShellWidgetClass; n++;
  class_array[n].class = transientShellWidgetClass; n++;
  class_array[n].class = xmDialogShellWidgetClass; n++;

  class_array[n].class = xmSashWidgetClass; n++;          /* primitive */
  class_array[n].class = xmTearOffButtonWidgetClass; n++; /* pushbutton */

  class_array[n].class = xmDisplayClass; n++;              /* appshell */
  class_array[n].class = xmScreenClass; n++;               /* core */
  class_array[n].class = xmDragContextClass; n++;          /* core */
  class_array[n].class = xmDragIconObjectClass; n++;       /* object */
  class_array[n].class = xmDropSiteManagerObjectClass; n++;/* object */
  class_array[n].class = xmDropTransferObjectClass; n++;   /* object */
  class_array[n].class = xmDragOverShellWidgetClass; n++;  /* vendors */

  class_array[n].class = xmGrabShellWidgetClass; n++;
  class_array[n].class = xmIconHeaderClass; n++;
  class_array[n].class = xmClipWindowWidgetClass; n++;
  class_array[n].class = xmContainerWidgetClass; n++;
  class_array[n].class = xmIconGadgetClass; n++;
  class_array[n].class = xmNotebookWidgetClass; n++;
  class_array[n].class = xmSpinBoxWidgetClass; n++;
  class_array[n].class = xmSimpleSpinBoxWidgetClass; n++;
  class_array[n].class = xmComboBoxWidgetClass; n++;
}

static WidgetClass GetClassPointer(String class)
{
    Cardinal i ;
    static Boolean first_time = True ;

    if (first_time) InitClassArray () ;
    first_time = False ;

    for (i = 0 ; i < XtNumber(class_array); i++) {
	if (strcmp(class, class_array[i].name) == 0)
	    return class_array[i].class ;
    }
    return NULL ;
}

static String buffer = NULL ;
static short rows = 0, columns = 0 ;

static void AddToBuffer(const char *fmt, ...)
{
    va_list args;
    static Cardinal curpos = 0 ;
    char tmp[256] ;
    Cardinal i, tmplen ;

    Va_start(args, fmt);

    (void)vsnprintf(tmp, sizeof tmp, fmt, args);
    tmplen = strlen(tmp) ;
    columns = (columns > tmplen)? columns : tmplen ;
    buffer = XtRealloc(buffer, curpos + tmplen + 1);
    for (i=0 ; tmp[i]; i++) {
	buffer[curpos++] = tmp[i] ;
	if (tmp[i] == '\n') rows ++ ;
    }
    buffer[curpos] = '\0' ;


    va_end(args);
}

static void GetPrintRes(String name)
{
    Cardinal i, j, k, num_resources, num_sec;
    XtResourceList resource_list ;
    XmSecondaryResourceData * res_sec_list ;
    WidgetClass class ;
    char buff_line[128];

    if ((class = GetClassPointer(name)) == NULL) {
	XtWarning("Cannot find specified class name");
	class = widgetClass ;
	name = "Widget" ;
    }

    XtInitializeWidgetClass (class) ;

    AddToBuffer("Fetching resources for widget %s:\n", name);
    AddToBuffer("=======================================\n");

    /* fetch Xt regular */
    XtGetResourceList(class, &resource_list, &num_resources);
    AddToBuffer("\nRegular Xt resources: %d\n",num_resources);
    AddToBuffer(  "------------------------\n");
    buff_line[0] = ' ';
    buff_line[1] = '\0';

    for (i = 0 ; i < num_resources; i++) {
	snprintf(buff_line + strlen(buff_line) - 1,
	         sizeof buff_line - strlen(buff_line),
	         "%-38s", resource_list[i].resource_name);

	if (i & 1) {
	    AddToBuffer(" %s\n", buff_line);
	    buff_line[0] = ' ';
	    buff_line[1] = '\0';
	}
    }

    if (!(i & 1))
	AddToBuffer(" %s\n", buff_line);
    XtFree((XtPointer)resource_list) ;

    /* fetch Xt constraint  */
    XtGetConstraintResourceList(class, &resource_list, &num_resources);
    if (num_resources) {
	AddToBuffer("\nConstraint Xt resources: %d\n",num_resources);
	AddToBuffer(  "------------------------\n");
	buff_line[0] = ' ';
	buff_line[1] = '\0';

	for (i = 0 ; i < num_resources; i++) {
	    snprintf(buff_line + strlen(buff_line) - 1,
	             sizeof buff_line - strlen(buff_line),
	             "%-38s", resource_list[i].resource_name);

	    if (i & 1) {
		AddToBuffer(" %s\n", buff_line);
		buff_line[0] = ' ';
		buff_line[1] = '\0';
	    }
	}

	if (!(i & 1))
	    AddToBuffer(" %s\n", buff_line);
	XtFree((XtPointer)resource_list) ;
    }

    /* fetch Motif second */
    if ((num_sec = XmGetSecondaryResourceData (class, &res_sec_list))) {
	AddToBuffer("\n\nMotif secondary blocks: %d\n", num_sec);
	AddToBuffer(    "--------------------------\n");

	for (i = 0; i < num_sec; i++) {
	    buff_line[0] = ' ';
	    buff_line[1] = '\0';
	    AddToBuffer("\nSecondary[%d] : %d resources\n\n", i,
	                res_sec_list[i]->num_resources);
	    for (j = 0 ; j < res_sec_list[i]->num_resources; j++) {
		snprintf(buff_line + strlen(buff_line) - 1,
		         sizeof buff_line - strlen(buff_line),
		         "%-38s", res_sec_list[i]->resources[j].resource_name);

		if (j & 1) {
		    AddToBuffer(" %s\n", buff_line);
		    buff_line[0] = ' ';
		    buff_line[1] = '\0';
	        }
	    }

	    if (!(j & 1))
		AddToBuffer(" %s\n", buff_line);
	    XtFree((XtPointer)res_sec_list[i]->resources);
	    XtFree((XtPointer)res_sec_list[i]);
	}

	XtFree((XtPointer)res_sec_list);
    }
}

static String fallbacks[] = {
    "*help_manager.helpFile: getsubres",
    NULL
};

#define APP_CLASS "XmdGetsubres"

int main(int argc, char *argv[])
{
    Widget toplevel, main_window, menu_bar, menu_pane, cascade,
           button, viewer ;
    XtAppContext app_context;
    String name ;
    Cardinal i ;
    Arg args[10] ;
    Cardinal n;


    /* Usage: getres class | getres All*/
    toplevel = XtAppInitialize(&app_context, APP_CLASS, NULL, 0,
			       &argc, argv, fallbacks, NULL, 0);

    /* get a default in no name provided */
    if (argc == 1) name = "Widget";
    else name = argv[1] ;

    /* Fill a string buffer with lines of text (this demo was originally
       designed with printf...) */
    if (strcmp(name, "All") == 0) {
	for (i = 0; i < XtNumber(class_array); i++) {
	    GetPrintRes (class_array[i].name);
	    AddToBuffer(
            "\n*******************************************************\n\n");
	}
    } else {
	GetPrintRes(name) ;
    }

    n = 0;
    main_window = XmCreateMainWindow (toplevel, "main_window", args, n);
    XtManageChild (main_window);

    /*	Create MenuBar in MainWindow.
     */
    n = 0;
    menu_bar = XmCreateMenuBar (main_window, "menu_bar", args, n);
    XtManageChild (menu_bar);


    /*	Create "File" PulldownMenu with a Quit button
     */
    n = 0;
    menu_pane = XmCreatePulldownMenu (menu_bar, "menu_pane", args, n);

    n = 0;
    button = XmCreatePushButton (menu_pane, "Quit", args, n);
    XtManageChild (button);
    XtAddCallback (button, XmNactivateCallback, QuitCB, NULL);

    n = 0;
    XtSetArg (args[n], XmNsubMenuId, menu_pane);  n++;
    cascade = XmCreateCascadeButton (menu_bar, "File", args, n);
    XtManageChild (cascade);


    /*	Create "Help" PulldownMenu with Help button.
     */
    n = 0;
    menu_pane = XmCreatePulldownMenu (menu_bar, "menu_pane", args, n);

    n = 0;
    button = XmCreatePushButton (menu_pane, "Overview", args, n);
    XtManageChild (button);
    XtAddCallback (button, XmNactivateCallback, HelpCB, NULL);

    n = 0;
    XtSetArg (args[n], XmNsubMenuId, menu_pane);  n++;
    cascade = XmCreateCascadeButton (menu_bar, "Help", args, n);
    XtManageChild (cascade);

    n = 0;
    XtSetArg (args[n], XmNmenuHelpWidget, cascade);  n++;
    XtSetValues (menu_bar, args, n);


    /* Create the viewer widget: here a text */
    n = 0;
    XtSetArg (args[n], XmNvalue, buffer);  n++;
    XtSetArg (args[n], XmNrows, rows+1);  n++;
    XtSetArg (args[n], XmNcolumns, columns-1);  n++;
    XtSetArg (args[n], XmNeditMode, XmMULTI_LINE_EDIT);  n++;
    if (strcmp(name, "All") != 0) {
	viewer = XmCreateText(main_window, "viewer", args, n);
    } else  {
	viewer = XmCreateScrolledText(main_window, "viewer", args, n);
    }
    XtManageChild(viewer);

    XtFree(buffer) ; /* Text has its own copy */

    XtRealizeWidget(toplevel);
    XtAppMainLoop(app_context);

    return 0;    /* make compiler happy */
}

/*-------------------------------------------------------------
**	QuitCB			- callback for quit button
*/
static void QuitCB(Widget w, XtPointer client, XtPointer callback)
{
    exit(EXIT_SUCCESS);
}

/*-------------------------------------------------------------
**	HelpCB			- callback for help button
*/
static void HelpCB(Widget w, XtPointer client, XtPointer callback)
{
	static Widget help_widget = NULL;

	if (!help_widget)
		help_widget = XmdCreateHelpDialog(w, "help_manager", NULL, 0);
	XtManageChild(help_widget);
}

